projects/extension/sql/idempotent/012-vectorizer-int.sql

```diff
@@ -485,59 +535,114 @@ begin
             end if;
-
-        elsif (TG_LEVEL = 'STATEMENT') then
-            if (TG_OP = 'TRUNCATE') then
-                execute format('truncate table %I.%I', '$TARGET_SCHEMA$', '$TARGET_TABLE$');
-                execute format('truncate table %I.%I', '$QUEUE_SCHEMA$', '$QUEUE_TABLE$');
+            
+            return null;
+        end;
+        $def$;
+    
+        -- Replace placeholders
+        _func_def := replace(_func_def, '$DELETE_STATEMENT$', _delete_statement);
+        _func_def := replace(_func_def, '$PK_CHANGE_CHECK$', _pk_change_check);
+        _func_def := replace(_func_def, '$QUEUE_SCHEMA$', quote_ident(queue_schema));
+        _func_def := replace(_func_def, '$QUEUE_TABLE$', quote_ident(queue_table));
+        _func_def := replace(_func_def, '$PK_COLUMNS$', _pk_columns);
+        _func_def := replace(_func_def, '$PK_VALUES$', _pk_values);
+        _func_def := replace(_func_def, '$TARGET_SCHEMA$', quote_ident(target_schema));
+        _func_def := replace(_func_def, '$TARGET_TABLE$', quote_ident(target_table));
+        _func_def := replace(_func_def, '$RELEVANT_COLUMNS_CHECK$', _relevant_columns_check);
+        _func_def := replace(_func_def, '$TRUNCATE_STATEMENT$', _truncate_statement);
+    else
+        _func_def := $def$
+        begin
+            if (TG_LEVEL = 'ROW') then
+                if (TG_OP = 'UPDATE') then
+                    if $RELEVANT_COLUMNS_CHECK$ then
+                        insert into $QUEUE_SCHEMA$.$QUEUE_TABLE$ ($PK_COLUMNS$)
+                        values ($PK_VALUES$);
+                    end if;
+                elseif (TG_OP = 'INSERT') then
+                    insert into $QUEUE_SCHEMA$.$QUEUE_TABLE$ ($PK_COLUMNS$)
+                    values ($PK_VALUES$);
+                end if;
             end if;
             return null;
-        end if;
-        
-        return null;
-    end;
-    $def$;
-
-    -- Replace placeholders
-    _func_def := replace(_func_def, '$DELETE_STATEMENT$', _delete_statement);
-    _func_def := replace(_func_def, '$PK_CHANGE_CHECK$', _pk_change_check);
-    _func_def := replace(_func_def, '$QUEUE_SCHEMA$', quote_ident(queue_schema));
-    _func_def := replace(_func_def, '$QUEUE_TABLE$', quote_ident(queue_table));
-    _func_def := replace(_func_def, '$PK_COLUMNS$', _pk_columns);
-    _func_def := replace(_func_def, '$PK_VALUES$', _pk_values);
-    _func_def := replace(_func_def, '$TARGET_SCHEMA$', quote_ident(target_schema));
-    _func_def := replace(_func_def, '$TARGET_TABLE$', quote_ident(target_table));
-
+        end;
+        $def$;
+        _func_def := replace(_func_def, '$RELEVANT_COLUMNS_CHECK$', _relevant_columns_check);
+        _func_def := replace(_func_def, '$QUEUE_SCHEMA$', quote_ident(queue_schema));
+        _func_def := replace(_func_def, '$QUEUE_TABLE$', quote_ident(queue_table));
+        _func_def := replace(_func_def, '$PK_COLUMNS$', _pk_columns);
+        _func_def := replace(_func_def, '$PK_VALUES$', _pk_values);
+    end if;
```

Comment:
There might be a somewhat smarter way to go about this, since the truncate and delete handling makes no sense for the same table vectorizer. But they share enough logic with the relevant column check that I didn't want to have two completely separate functions.